File: src/entry.c
/* Copyright (C) 2022 Daniel Forde - All Rights Reserved
 * You may use, distribute and modify this code under the
 * terms of the MIT license.
 *
 * You should have received a copy of the MIT license with this file.
 * If not, please write to: <daniel.forde001 at gmail dot com>,
 * or visit: https://github.com/deforde/tmenu
 */
#include "entry.h"

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

static bool createRealpath(char (*realpath)[PATH_MAX], const char *branch,
                           const char *leaf);
static bool isRelevantFiletype(char (*realpath)[PATH_MAX]);

Entry *entryCreate(const Allocator *allocator, const char *s) {
  size_t nm_len = strlen(s) + 1;
  size_t sz = sizeof(Entry) + nm_len;
  Entry *e = allocator->alloc(sz);
  if (e) {
    strncpy(e->path, s, nm_len);
    e->name = e->path;
    char *sep = strrchr(e->path, '/');
    if (sep) {
      e->name = ++sep;
      assert(e->name); // TODO: add proper error-checking
    }
    assert(e->path[nm_len - 1] == 0); // TODO: add proper error-checking
  }
  return e;
}

void entryDestroy(const Allocator *allocator, Entry *e) { allocator->free(e); }

bool createRealpath(char (*realpath)[PATH_MAX], const char *branch,
                    const char *leaf) {
  int r = snprintf(*realpath, sizeof(*realpath), "%s/%s", branch, leaf);
  if (r == -1) {
    perror("snprintf");
    return false;
  }
  if (r >= (int)sizeof(*realpath)) {
    printf("snprintf: destination buffer too small (actual size: %zu, "
           "expected: %i)\n",
           sizeof(*realpath), r);
    return false;
  }
  return true;
}

bool isRelevantFiletype(char (*realpath)[PATH_MAX]) {
  struct stat sb;
  if (lstat(*realpath, &sb) == -1) {
    perror("lstat");
    return false; // TODO: Error here?
  }
  if (sb.st_mode & S_IFMT) {
    memset(&sb, 0, sizeof(sb));
    if (stat(*realpath, &sb) == -1) {
      perror("stat");
      return false; // TODO: Error here?
    }
  }
  return sb.st_mode & S_IXUSR;
}

EntryList entrylistInit(const Allocator *allocator) {
  EntryList entries = {
      .head = NULL,
      .tail = NULL,
  };

  char *path = getenv("PATH");
  if (path == NULL) {
    printf("Failed to get the environment variable: 'PATH'\n");
    return entries;
  }

  char *saveptr = NULL;
  for (char *str = path;; str = NULL) {
    char *tok = strtok_r(str, ":", &saveptr);
    if (tok == NULL) {
      break;
    }

    struct dirent **namelist = NULL;
    int n = scandir(tok, &namelist, NULL, alphasort);
    if (n == -1) {
      // perror("scandir");
      continue;
    }

    while (n--) {
      struct dirent *dent = namelist[n];
      if (dent->d_type == DT_REG || dent->d_type == DT_LNK) {
        char realpath[PATH_MAX] = {0};
        if (!createRealpath(&realpath, tok, dent->d_name)) {
          goto err;
        }
        if (isRelevantFiletype(&realpath)) {
          // printf("%s\n", dent->d_name);
          Entry *e = entryCreate(allocator, realpath);
          if (e == NULL) {
            printf("Failed to allocate memory for entry\n");
            goto err;
          }
          if (!entrylistAppendUnique(&entries, e)) {
            entryDestroy(allocator, e);
          }
        }
      }

    err:
      free(dent);
    }

    free(namelist);
  }

  return entries;
}

void entrylistAppend(EntryList *l, Entry *e) {
  e->prev = l->tail;
  if (!l->head) {
    l->head = e;
  }
  if (l->tail) {
    l->tail->next = e;
  }
  l->tail = e;
  l->len++;
}

void entrylistExtend(EntryList *l, EntryList m) {
  entrylistAppend(l, m.head);
  l->tail = m.tail;
  l->len += m.len - 1;
}

bool entrylistAppendUnique(EntryList *l, Entry *e) {
  for (Entry *p = l->head; p; p = p->next) {
    if (strlen(p->name) == strlen(e->name) && strcmp(p->name, e->name) == 0) {
      return false;
    }
  }
  entrylistAppend(l, e);
  return true;
}

void entrylistRemove(EntryList *l, Entry *e) {
  Entry *prev = e->prev;
  Entry *next = e->next;
  if (prev) {
    prev->next = next;
  }
  if (next) {
    next->prev = prev;
  }
  e->prev = NULL;
  e->next = NULL;
  if (e == l->head) {
    l->head = next;
  }
  if (e == l->tail) {
    l->tail = prev;
  }
  l->len--;
}

void entrylistDestroy(const Allocator *allocator, EntryList *l) {
  for (Entry *e = l->head; e;) {
    Entry *tmp = e->next;
    entryDestroy(allocator, e);
    e = tmp;
  }
  l->head = NULL;
  l->tail = NULL;
  l->len = 0;
}

void entrylistPrint(EntryList l) {
  for (Entry *e = l.head; e; e = e->next) {
    printf("%s\n", e->name);
  }
}

void entrylistFilter(EntryList *l, EntryList *fout, const char *s) {
  if (s[0] == 0) {
    return;
  }
  for (Entry *e = l->head; e;) {
    if (!strstr(e->name, s)) {
      Entry *tmp = e->next;
      entrylistRemove(l, e);
      entrylistAppend(fout, e);
      e = tmp;
      continue;
    }
    e = e->next;
  }
}

void entrylistClear(EntryList *l) {
  l->head = NULL;
  l->tail = NULL;
  l->len = 0;
}

File: src/entry.zig
// Copyright (C) 2022 Daniel Forde - All Rights Reserved
// You may use, distribute and modify this code under the
// terms of the MIT license.
//
// You should have received a copy of the MIT license with this file.
// If not, please write to: <daniel.forde001 at gmail dot com>,
// or visit: https://github.com/deforde/tmenu

const std = @import("std");
const Allocator = std.mem.Allocator;

pub const Entry = struct {
    prev: ?*Entry = null,
    next: ?*Entry = null,
    name: ?[:0]u8 = null,
    path: ?[:0]u8 = null,
    allocator: *const Allocator,

    pub fn create(allocator: *const Allocator, s: []const u8) !*Entry {
        var e = try allocator.create(Entry);
        e.* = Entry{ .allocator = allocator };
        e.path = try allocator.allocSentinel(u8, s.len, 0);
        std.mem.copy(u8, e.path.?, s);
        const pos = std.mem.lastIndexOf(u8, e.path.?, "/");
        if (pos != null) {
            e.name = e.path.?[pos.? + 1 ..];
        } else {
            e.name = e.path;
        }
        return e;
    }

    pub fn destroy(self: *Entry) void {
        if (self.path != null) {
            self.allocator.free(self.path.?);
        }
        self.allocator.destroy(self);
    }
};

pub const EntryList = struct {
    head: ?*Entry = null,
    tail: ?*Entry = null,
    len: usize = 0,

    pub fn append(self: *EntryList, e: *Entry) void {
        e.prev = self.tail;
        if (self.head == null) {
            self.head = e;
        }
        if (self.tail != null) {
            self.tail.?.next = e;
        }
        self.tail = e;
        self.len += 1;
    }

    pub fn remove(self: *EntryList, e: *Entry) void {
        var prev = e.prev;
        var next = e.next;
        if (prev != null) {
            prev.?.next = next;
        }
        if (next != null) {
            next.?.prev = prev;
        }
        e.prev = null;
        e.next = null;
        if (e == self.head) {
            self.head = next;
        }
        if (e == self.tail) {
            self.tail = prev;
        }
        self.len -= 1;
    }

    pub fn clear(self: *EntryList) void {
        self.head = null;
        self.tail = null;
        self.len = 0;
    }

    pub fn destroy(self: *EntryList) void {
        var e = self.head;
        while (e != null) {
            const tmp = e.?.next;
            e.?.destroy();
            e = tmp;
        }
        self.clear();
    }

    pub fn print(self: *EntryList) void {
        var e = self.head;
        while (e != null) : (e = e.?.next) {
            std.debug.print("{s}\n", .{e.?.name.?});
        }
    }

    pub fn appendUnique(self: *EntryList, e: *Entry) bool {
        var ext = self.head;
        while (ext != null) : (ext = ext.?.next) {
            if (std.mem.eql(u8, ext.?.name.?, e.name.?)) {
                return false;
            }
        }
        self.append(e);
        return true;
    }

    pub fn extend(self: *EntryList, l: EntryList) void {
        if (l.head == null or l.tail == null) {
            return;
        }
        self.append(l.head.?);
        self.tail = l.tail;
        self.len += l.len - 1;
    }

    pub fn filter(self: *EntryList, fout: *EntryList, s: []const u8) void {
        if (s.len == 0) {
            return;
        }
        var e = self.head;
        while (e != null) {
            const idx = std.mem.indexOf(u8, e.?.name.?, s);
            if (idx == null) {
                const tmp = e.?.next;
                self.remove(e.?);
                fout.append(e.?);
                e = tmp;
                continue;
            }
            e = e.?.next;
        }
    }

    pub fn create(allocator: *const Allocator) !EntryList {
        var l = EntryList{};

        const env_path = std.os.getenv("PATH").?;

        var dir_paths = std.mem.tokenize(u8, env_path, ":");
        while (dir_paths.next()) |dir_path| {
            var dir = std.fs.openIterableDirAbsolute(dir_path, .{}) catch continue;
            defer dir.close();
            var it = dir.iterate();
            while (try it.next()) |dir_entry| {
                if ((dir_entry.kind == std.fs.File.Kind.File or dir_entry.kind == std.fs.File.Kind.SymLink) and !std.mem.eql(u8, dir_entry.name, "tmenu")) {
                    const realpath = try std.fmt.allocPrint(allocator.*, "{s}/{s}", .{ dir_path, dir_entry.name });
                    defer allocator.free(realpath);
                    var file = std.fs.openFileAbsolute(realpath, .{}) catch continue;
                    defer file.close();
                    var fstat = try std.fs.File.stat(file);
                    if (fstat.mode & std.os.linux.S.IXUSR != 0) {
                        var e = try Entry.create(allocator, realpath);
                        if (!l.appendUnique(e)) {
                            e.destroy();
                        }
                    }
                }
            }
        }

        return l;
    }
};
